Interfacing IpOpt from Crocoddyl

PepMS opened the pull request:
This is the PR following #1019 

As requested by @cmastalli , I've included a solver class that solves a Crocoddyl problem with IpOpt. Also, I've added the class to the existing unit tests for solvers.

Some known issues of this initial development:
-  The Ipopt dependency is not optional. Even though it has been introduced as such in the CMake (https://github.com/PepMS/crocoddyl/blob/8aeb4d354d58b22236154b553291816108b9ff7a/CMakeLists.txt#L71), files including ipopt dependencies should be excluded in case ipopt is not found. I would need some guidance to do so.
-  LQR unit tests do not pass, since I implemented the solver assuming the number of controls would be the same for all nodes in the OCP. I will implement this in the coming days.

cmastalli commented on bindings/python/crocoddyl/core/core.cpp:
```diff
@@ -60,6 +60,7 @@ void exposeCore() {
   exposeSolverBoxQP();
   exposeSolverBoxDDP();
   exposeSolverBoxFDDP();
+  exposeSolverIpOpt();
```
```suggestion
#ifdef CROCODDYL_WITH_IPOPT
  exposeSolverIpopt();
#endif  // CROCODDYL_WITH_IPOPT
```

cmastalli commented on CMakeLists.txt:
```diff
@@ -68,6 +68,7 @@ SET(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake/find-external/CppAD/" ${CMAKE_M
 ADD_PROJECT_DEPENDENCY(example-robot-data 3.7.0 REQUIRED PKG_CONFIG_REQUIRES "example-robot-data >= 3.7.0")
 ADD_OPTIONAL_DEPENDENCY("scipy")
+ADD_OPTIONAL_DEPENDENCY(ipopt)
 
```
To enable Ipopt as an optional library, we need to add a definition that we'll pass to the library.
The code would be something like this

```
IF(IPOPT_FOUND)
  ADD_DEFINITIONS(-DCROCODDYL_WITH_IPOPT)
ENDIF()
```

Then, we'll use this macro (in the c++ files) to include or not the Ipopt files. 
See this example here: https://github.com/loco-3d/crocoddyl/pull/1020/files#r777798630

I would try to mark all the places where we need to use this macro.

cmastalli commented on include/crocoddyl/core/solvers/ipopt/ipopt.hpp:
```diff
@@ -0,0 +1,30 @@
+
+#include "crocoddyl/core/solver-base.hpp"
+#include "crocoddyl/core/solvers/ipopt/multiple-shooting-nlp.hpp"
+namespace crocoddyl {
+class SolverIpOpt : public SolverAbstract {
```
As mentioned above, please rename this function to `SolverIpopt`.
Also add the line between the header and namespace (sorry for this picky comment ;))

cmastalli commented on include/crocoddyl/core/solvers/ipopt/ipopt.hpp:
```diff
@@ -0,0 +1,30 @@
+#ifndef CROCODDYL_CORE_SOLVERS_IPOPT_IPOPT_HPP_
+#define CROCODDYL_CORE_SOLVERS_IPOPT_IPOPT_HPP_
+
+#include "coin-or/IpIpoptApplication.hpp"
```
```suggestion
#include <coin-or/IpIpoptApplication.hpp>
```

This is part of our code style.

cmastalli commented on include/crocoddyl/core/solvers/ipopt/ipopt.hpp:
```diff
@@ -0,0 +1,30 @@
+#include "coin-or/IpIpoptApplication.hpp"
+
+#include "crocoddyl/core/solver-base.hpp"
+#include "crocoddyl/core/solvers/ipopt/multiple-shooting-nlp.hpp"
```
Renamed this file to ipopt-iface.hpp, i.e.,
```suggestion
#include "crocoddyl/core/solvers/ipopt/ipopt-iface.hpp"
```

cmastalli commented on include/crocoddyl/core/solvers/ipopt/ipopt.hpp:
```diff
@@ -0,0 +1,30 @@
+             const bool is_feasible = false, const double regInit = 1e-9);
+
+ private:
+  Ipopt::SmartPtr<MultipleShootingNlp> ms_nlp_;
```
Renamed this class to `IpoptInterface`, i.e.,
```suggestion
  Ipopt::SmartPtr<IpoptInterface> ipopt_iface_;
```

cmastalli commented on include/crocoddyl/core/solvers/ipopt/ipopt.hpp:
```diff
@@ -0,0 +1,30 @@
+
+ private:
+  Ipopt::SmartPtr<MultipleShootingNlp> ms_nlp_;
+  Ipopt::SmartPtr<Ipopt::IpoptApplication> app_;
+  Ipopt::ApplicationReturnStatus status_;
```
Renamed these variables as suggested below
```suggestion
  Ipopt::SmartPtr<Ipopt::IpoptApplication> ipopt_app_;
  Ipopt::ApplicationReturnStatus ipopt_status_;
```

cmastalli commented on include/crocoddyl/core/solvers/ipopt/multiple-shooting-nlp.hpp:
```diff
@@ -0,0 +1,178 @@
+#ifndef __CROCODDYL_CORE_SOLVERS_IPOPT_MULTIPLE_SHOOTING_NLP_HPP__
```
Define the macro as in the `ipopt.hpp`

cmastalli commented on bindings/python/crocoddyl/core/solvers/ipopt/ipopt.cpp:
```diff
@@ -0,0 +1,33 @@
+#include "python/crocoddyl/core/core.hpp"
+#include "crocoddyl/core/solvers/ipopt/ipopt.hpp"
+
```
Do not create an unnecessary sub-folder.
Please move this file to
`bindings/python/crocoddyl/core/solvers/ipopt.cpp`

cmastalli commented on bindings/python/crocoddyl/core/solvers/ipopt/ipopt.cpp:
```diff
@@ -0,0 +1,33 @@
+#include "python/crocoddyl/core/core.hpp"
```
```suggestion
#ifdef CROCODDYL_WITH_IPOPT
#include "python/crocoddyl/core/core.hpp"
```

cmastalli commented on bindings/python/crocoddyl/core/solvers/ipopt/ipopt.cpp:
```diff
@@ -0,0 +1,33 @@
+}
+
+}  // namespace python
+}  // namespace crocoddyl
```
```suggestion
#endif  // CROCODDYL_WITH_IPOPT
```

cmastalli commented on include/crocoddyl/core/solvers/ipopt/multiple-shooting-nlp.hpp:
```diff
@@ -0,0 +1,178 @@
+#define __CROCODDYL_CORE_SOLVERS_IPOPT_MULTIPLE_SHOOTING_NLP_HPP__
+
+#include "coin-or/IpTNLP.hpp"
+
```
This file doesn't follow our code format.
Please follow the instructions (carefully) in the wiki to automatically format the code.

PepMS replied:
I forgot to format this one. Done

cmastalli commented on src/core/solvers/ipopt/ipopt.cpp:
```diff
@@ -0,0 +1,48 @@
+  assert(init_xs.size() == ms_nlp_->get_problem()->get_T() + 1);
+  assert(init_us.size() == ms_nlp_->get_problem()->get_T());
+
+  if (init_xs != DEFAULT_VECTOR) {
+    ms_nlp_->set_xs(init_xs);
+  }
+  if (init_us != DEFAULT_VECTOR) {
+    ms_nlp_->set_us(init_us);
+  }
```
Use  `setCandidate`, i.e.,
```suggestion
setCandidate(init_xs, init_us, is_feasible);
ipopt_iface_->set_xs(xs_);
ipopt_iface_->set_us(us_);
```

cmastalli commented on src/core/solvers/ipopt/ipopt.cpp:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  app_->Options()->SetIntegerValue("max_iter", maxiter);
+  Ipopt::ApplicationReturnStatus status = app_->OptimizeTNLP(ms_nlp_);
```
Why are you not using `status_`?

PepMS replied:
True. Fixed it

cmastalli commented on include/crocoddyl/core/solvers/ipopt/multiple-shooting-nlp.hpp:
```diff
@@ -0,0 +1,178 @@
+#ifndef __CROCODDYL_CORE_SOLVERS_IPOPT_MULTIPLE_SHOOTING_NLP_HPP__
+#define __CROCODDYL_CORE_SOLVERS_IPOPT_MULTIPLE_SHOOTING_NLP_HPP__
+
+#include "coin-or/IpTNLP.hpp"
```
```suggestion
#include <coin-or/IpTNLP.hpp>
```

cmastalli commented on include/crocoddyl/core/solvers/ipopt/multiple-shooting-nlp.hpp:
```diff
@@ -0,0 +1,178 @@
+
+    private:
+    boost::shared_ptr<crocoddyl::ShootingProblem> problem_;
+    boost::shared_ptr<crocoddyl::StateAbstract>   state_;
```
You don't need to store another shared pointer of the state.
You could retrieve it from `problem_`.

PepMS replied:
How could I do that? I don't see any getter to retrieve the state from problem. You mean through the `runningModels`?

What is the problem of having another shared pointer of the state? Having this variable inside the class simplifies things when calling to integrate or its Jacobians

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+      state_(problem_->get_runningModels()[0]->get_state()),
+      nx_(problem_->get_nx()),
+      ndx_(problem_->get_ndx()),
+      nu_(problem_->get_nu_max()),
```
Let's do not allocate the maximum nu.
This got deprecated recently (see `SolverDDP` in devel branch for more details).

cmastalli replied:
This implies that you need to modify a few lines of code, I'll let you to figure out them and later I could make comments on it.

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+{
+    // here, the n and m we gave IPOPT in get_nlp_info are passed back to us.
+    // If desired, we could assert to make sure they are what we think they are.
+    assert(n == nvar_);
+    assert(m == nconst_);
```
Let's use ``assert_pretty` as it triggers issues in the Python code as well (without the need to compile in Debug mode)

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+    for (size_t i = 0; i < T_; i++) {
+        // Running state bounds
+        for (size_t j = 0; j < ndx_; j++) {
+            x_l[i * (ndx_ + nu_) + j] = -2e19;
+            x_u[i * (ndx_ + nu_) + j] = 2e19;
```
You should use `std::numeric_limits` to get the maximum value.

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+        }
+        // Control bounds
+        for (size_t j = 0; j < nu_; j++) {
+            x_l[i * (ndx_ + nu_) + ndx_ + j] = problem_->get_runningModels()[i]->get_has_control_limits()
```
Simplify the code as
```suggestion
    const boost::shared_ptr<ActionModelAbstract>& model = problem_->get_runningModels()[i];
    const boost::shared_ptr<ActionDataAbstract>& data = problem_->get_runningDatas()[i];
            x_l[i * (ndx_ + nu_) + ndx_ + j] = model->get_has_control_limits()
```

Do this in other case. I won't mark them.

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+    // Running costs
+    for (size_t i = 0; i < T_; i++) {
+        data_.dx      = Eigen::VectorXd::Map(x + i * (ndx_ + nu_), ndx_);
+        data_.control = Eigen::VectorXd::Map(x + i * (ndx_ + nu_) + ndx_, nu_);
```
Rename this data as `data_.u`

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+
+        state_->integrate(xs_[i], data_.dx, data_.x);
+
+        problem_->get_runningModels()[i]->calc(problem_->get_runningDatas()[i], data_.x, data_.control);
```
It is better to run firs the integration operators over the entire trajectory (and store then in `xs_guess_`).
Then, you call `problem_->calc(xs_guess_, us_guess_)

This approach will run the different computations in parallel, if the user has enabled it.

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+        state_->integrate(xs_[i], data_.dx, data_.x);
+
+        state_->Jintegrate(xs_[i], data_.dx, data_.Jsum_x, data_.Jsum_dx, crocoddyl::second, crocoddyl::setto);
+        problem_->get_runningModels()[i]->calcDiff(problem_->get_runningDatas()[i], data_.x, data_.control);
```
Same approach here.
I also presumed that `xs_guess_` and `us_guess` is the same stored in `eval_f`.
Therefore, you could avoid this step and directly call `problem_->calcDiff()`

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+
+        state_->Jintegrate(xs_[i], data_.dx, data_.Jsum_x, data_.Jsum_dx, crocoddyl::second, crocoddyl::setto);
+        problem_->get_runningModels()[i]->calcDiff(problem_->get_runningDatas()[i], data_.x, data_.control);
+        data_.Ldx = data_.Jsum_dx.transpose() * problem_->get_runningDatas()[i]->Lx;
```
If I understand correctly this, then it would be better (more efficient) to use `JintegrateTransport`.
See this example: https://github.com/loco-3d/crocoddyl/blob/master/include/crocoddyl/core/integrator/euler.hxx#L113-L115

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+        state_->integrate(xs_[i], data_.dx, data_.x);
+        state_->integrate(xs_[i + 1], data_.dxnext, data_.xnext);
+
+        problem_->get_runningModels()[i]->calc(problem_->get_runningDatas()[i], data_.x, data_.control);
```
Are you sure we need to run `calc`?
I suspect this is not necessary as we run it for the cost computation.

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+            state_->integrate(xs_[idx_block], data_.dx, data_.x);
+            state_->integrate(xs_[idx_block + 1], data_.dxnext, data_.xnext);
+
+            problem_->get_runningModels()[idx_block]->calc(problem_->get_runningDatas()[idx_block], data_.x,
+                                                           data_.control);
+            problem_->get_runningModels()[idx_block]->calcDiff(problem_->get_runningDatas()[idx_block], data_.x,
+                                                               data_.control);
```
Same here.
We might not need to run `calc` and `calcDiff`.

Btw, you haven't run `calc` in the computation of cost derivatives!

cmastalli commented on src/core/solvers/ipopt/multiple-shooting-nlp.cpp:
```diff
@@ -0,0 +1,589 @@
+                                 Ipopt::Index        *jCol,
+                                 Ipopt::Number       *values)
+{
+    assert(n == nvar_);
+    assert(m == nconst_);
```
Use assert_pretty in all the cases

cmastalli reviewed (commented):
@PepMS --- there are much more things to be commented, however, I decided to do a quick feedback.

Could you let me know when I could review again the code?

(Note that I provided a solution for your first bulletpoint, I will provide more details about the second one in the next review)

PepMS commented:
@cmastalli During this week I'll be able to spend a couple of hours each day since I'm on "holiday". It is difficult to say how much time I can spend on this from next week on, since I'll go back to the lab.

cmastalli commented:
@nim65s -- can you help me to include Ipopt in gitlab-ci?
@wxmerkt -- could you the same in ROS-CI?

This seems to be viable via Conda: https://anaconda.org/conda-forge/ipopt.
And enabling this feature before merging this PR would be of tremendous help for maintenance purpose.

cmastalli commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+    const std::size_t nui = models[t]->get_nu();
+
+    for (std::size_t j = 0; j < ndxi; j++) {
+      x[ixu + j] = 0;
```
It seems to me that we should use the deltas xs. You should compute as xs_[i] (-) 0 = diff(0, xs_[i])

PepMS replied:
I've implemented the following: $\mathbf{x}(0) \ominus (\mathbf{x}^0_k \oplus \mathbf{\Delta x}_k) = \mathbf{0}$

cmastalli commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+    model->get_state()->integrate(xs_[t], datas_[t]->dx, datas_[t]->x);
+    model->get_state()->Jintegrate(xs_[t], datas_[t]->dx, datas_[t]->Jsum_dx, datas_[t]->Jsum_dx, second, setto);
+
+    model->calc(data, datas_[t]->x, datas_[t]->u);
```
Are you sure we need to run `calc`? it seems improbable to me that Ipopt needs this.

PepMS replied:
It does need this call. We need this calc here since this function is the first to be called to compute the search direction.

Removing this call the solver exits without running a single iteration.

cmastalli commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+
+    model->calc(data, datas_[t]->x, datas_[t]->u);
+    model->calcDiff(data, datas_[t]->x, datas_[t]->u);
+    datas_[t]->Ldx = datas_[t]->Jsum_dx.transpose() * data->Lx;
```
Use `noalias()`

cmastalli commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+    model->get_state()->integrate(xs_[t], datas_[t]->dx, datas_[t]->x);
+    model_next->get_state()->integrate(xs_[t + 1], datas_[t + 1]->dx, datas_[t + 1]->x);
+
+    model->calc(data, datas_[t]->x, datas_[t]->u);
```
I do believe that Ipopt should call always eval_f or eval_g first. If so, it means that we need to call `calc` in one of these functions.

Can you check this?

PepMS replied:
calc() is called first at `eval_grad_f`. However, I've only tested this with the few examples I have.

If we want to optimize properly the calls to `calc` and `calcDiff` I think we should do it by considering the IPOPT algorithm. Otherwise, we might be skipping some case.

I'd take a safe path in this case and leave the calls as they are. Unfortunately, at this moment, I do not have the time to do such optimization.

cmastalli commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+
+      model->calcDiff(data, datas_[t]->x, datas_[t]->u);
+
+      model_next->get_state()->Jintegrate(xs_[t + 1], datas_[t + 1]->dx, datas_[t + 1]->Jsum_dx,
```
Let's call Jsum_dx as Jint_x

cmastalli commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+
+      model->calcDiff(data, datas_[t]->x, datas_[t]->u);
+
+      model_next->get_state()->Jintegrate(xs_[t + 1], datas_[t + 1]->dx, datas_[t + 1]->Jsum_dx,
+                                          datas_[t + 1]->Jsum_dx, second, setto);  // datas_[t]->Jsum_dxnext == eq. 81
+      model->get_state()->Jdiff(data->xnext, datas_[t + 1]->x, datas_[t]->Jdiff_x, datas_[t + 1]->Jdiff_x,
+                                both);  // datas_[t+1]->Jdiff_x == eq. 83, datas_[t]->Jdiff_x == eq.82
+      model->get_state()->Jintegrate(xs_[t], datas_[t]->dx, datas_[t]->Jsum_dx, datas_[t]->Jsum_dx, second,
+                                     setto);  // datas_[t]->Jsum_dx == eq. 81
```
If we create data for next values, then we could parallelise this routine.
Additionally, there are other routines that are possible to parallelise now.

PepMS replied:
Optimization related

cmastalli commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+      datas_[t]->dx = Eigen::VectorXd::Map(x + ixu, ndxi);
+      datas_[t]->u = Eigen::VectorXd::Map(x + ixu + ndxi, nui);
+
+      model->get_state()->integrate(xs_[t], datas_[t]->dx, datas_[t]->x);
```
We shouldn't need to perform this integration again, right?
This should be done in the function that compute the cost or constraints.

Most likely this comment can be applied in other places.

PepMS replied:
Same as other optimization related comments. Without knowing what is called first we cannot do this kind of optimizations.

cmastalli commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+
+      datas_[T_]->dx = Eigen::VectorXd::Map(x + ixu, ndxi);
+      model->get_state()->integrate(xs_[T_], datas_[T_]->dx, datas_[T_]->x);
+      model->calc(data, datas_[T_]->x);
```
I don't think we need calc, right?

PepMS replied:
As I said in the previous comment, I'd prefer to leave this kind of code optimization for later.

I would like to di it in the proper way, i.e., by knowing the Ipopt algorithm and removing del calcs with more certainty.

In any case, I don't think we want Ipopt as a fast solver but more like a reliable and solver. Not saying that being faster is not important, but less important here.

cmastalli commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+      datas_[T_]->dx = Eigen::VectorXd::Map(x + ixu, ndxi);
+      model->get_state()->integrate(xs_[T_], datas_[T_]->dx, datas_[T_]->x);
+      model->calc(data, datas_[T_]->x);
+      model->calcDiff(data, datas_[T_]->x);
```
And calcDiff to be revise as well. Most likely Ipopt computes the Jacobians first. So this might be called in one of those functions.

PepMS replied:
Same as previous comments

cmastalli commented on unittest/factory/solver.cpp:
```diff
@@ -12,6 +12,7 @@
 #include "crocoddyl/core/solvers/fddp.hpp"
 #include "crocoddyl/core/solvers/box-ddp.hpp"
 #include "crocoddyl/core/solvers/box-fddp.hpp"
+#include "crocoddyl/core/solvers/ipopt.hpp"
```
Check if we compile Crocoddyl with Ipopt support.

cmastalli commented on unittest/factory/solver.cpp:
```diff
@@ -36,6 +37,9 @@ std::ostream& operator<<(std::ostream& os, SolverTypes::Type type) {
       os << "SolverBoxFDDP";
       break;
+    case SolverTypes::SolverIpopt:
+      os << "SolverIpopt";
```
Do not define this case if we haven't installed Ipopt.

cmastalli commented on unittest/factory/solver.cpp:
```diff
@@ -83,6 +87,9 @@ boost::shared_ptr<crocoddyl::SolverAbstract> SolverFactory::create(SolverTypes::
     case SolverTypes::SolverBoxFDDP:
       solver = boost::make_shared<crocoddyl::SolverBoxFDDP>(problem);
       break;
+    case SolverTypes::SolverIpopt:
```
Same here

cmastalli commented on unittest/factory/solver.hpp:
```diff
@@ -18,7 +18,7 @@ namespace crocoddyl {
 namespace unittest {
 
 struct SolverTypes {
-  enum Type { SolverKKT, SolverDDP, SolverFDDP, SolverBoxDDP, SolverBoxFDDP, NbSolverTypes };
```
Same here

cmastalli reviewed (changes requested):
Thanks for pushing this feature forward.
The PR is approaching to be merge with the latest status.

Could you share logs and results with Ipopt?

After you tackles my comments, I would like to run it myself. And eventually do "last minutes" changes.

cmastalli commented:
Please feel free to mark as resolved any comment that you have addressed.

nim65s commented:
As long as ipopt is an optional dependency, existing CI should succeed without ipopt, by just not building/testing ipopt related stuff.

I could add iptop to gitlab-ci, but once this will be done, there will be no more tests for "crocoddyl-without-ipopt" on my side, and this is currently broken.

The first issue according to gitlab-ci is:
```
src/core/solvers/ipopt.cpp:5:1: error: 'SolverIpopt' does not name a type
5 | SolverIpopt::SolverIpopt(boost::shared_ptr<crocoddyl::ShootingProblem> problem)
```

Do you need any help to fix that ?

Otherwise, an alternative would be to require ipopt ;)

cmastalli commented:
> As long as ipopt is an optional dependency, existing CI should succeed without ipopt, by just not building/testing ipopt related stuff.
> 
> I could add iptop to gitlab-ci, but once this will be done, there will be no more tests for "crocoddyl-without-ipopt" on my side, and this is currently broken.

I think it is convenient to keep Ipopt as an optional dependency. But it would be nice to have an extra job that tests Crocoddyl with Ipopt support. Is this possible?



> The first issue according to gitlab-ci is:
> 
> ```
> src/core/solvers/ipopt.cpp:5:1: error: 'SolverIpopt' does not name a type
> 5 | SolverIpopt::SolverIpopt(boost::shared_ptr<crocoddyl::ShootingProblem> problem)
> ```
> 
> Do you need any help to fix that ?

I do believe this is happening as we are not using the macro to disable Ipopt stuff in the unit-tests. @PepMS might be able to handle it, right?

> Otherwise, an alternative would be to require ipopt ;)

As said, let's keep Ipopt as an optional dependency. Do you agree with this decision?

nim65s commented:
> > As long as ipopt is an optional dependency, existing CI should succeed without ipopt, by just not building/testing ipopt related stuff.
> > I could add iptop to gitlab-ci, but once this will be done, there will be no more tests for "crocoddyl-without-ipopt" on my side, and this is currently broken.
> 
> I think it is convenient to keep Ipopt as an optional dependency. But it would be nice to have an extra job that tests Crocoddyl with Ipopt support. Is this possible?

No. Robotpkg builds will be with or without ipopt, this won't be an available option as it can be in the sources. 
I think it should be with ipopt.

cmastalli commented:
> No. Robotpkg builds will be with or without ipopt, this won't be an available option as it can be in the sources.
> I think it should be with ipopt.

Ok, I see. I was talking about something slightly different: CI job with Ipopt.
Yes, I agree with you. It is a good idea to release Crocoddyl binaries with Ipopt support.

PepMS commented on src/core/solvers/ipopt/ipopt-iface.cpp:
```diff
@@ -0,0 +1,665 @@
+
+#include "crocoddyl/core/solvers/ipopt/ipopt-iface.hpp"
+
+#ifdef __GNUC__
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#endif
```
This is legacy code from the Ipopt example. I've removed them.

cmastalli commented on include/crocoddyl/core/solvers/ipopt/ipopt-iface.hpp:
```diff
@@ -0,0 +1,143 @@
+
+struct IpoptInterfaceData;
+
+class IpoptInterface : public Ipopt::TNLP {
```
At the moment, I do not see any doxygen-related code. Are you having this issue with local changes that introduce the doxygen code?

Please also check that you have installed doxygen in your PC. This is an optional dependency in Crocoddyl.

PepMS commented on src/core/solvers/ipopt.cpp:
```diff
@@ -0,0 +1,51 @@
+
+SolverIpopt::SolverIpopt(boost::shared_ptr<crocoddyl::ShootingProblem> problem)
+    : SolverAbstract(problem), ipopt_iface_(new IpoptInterface(problem)), ipopt_app_(IpoptApplicationFactory()) {
+  ipopt_app_->Options()->SetNumericValue("tol", 3.82e-6);
```
Yes, this is the th_stop value. I will change that. However, the `set_th_stop` method is not virtual. How should I address this?

PepMS commented on src/core/solvers/ipopt.cpp:
```diff
@@ -0,0 +1,51 @@
+  ipopt_iface_->set_xs(xs_);
+  ipopt_iface_->set_us(us_);
+
+  ipopt_app_->Options()->SetIntegerValue("max_iter", maxiter);
```
To which method are you referring to? In `solver-base.hpp` I cannot see any `set_maxiter()` method.

PepMS commented:
@cmastalli I addressed all the comments (I think) except those related to code optimization. These are mainly related to the need (or not) to call `calc()` and `calcDiff()` inside the evaluation methods. As I already said in some of my comments, I think that the decision of calling or not these methods has to be made considering the Ipopt algorithm described in its corresponding paper.
I'm not comfortable doing otherwise. Unfortunately, right now I don't have the time to do it.

Besides, I think that Ipopt should bring reliability rather than fast (we know it is not gonna be fast). I'm not saying optimizing the code is not important, but it can be left for a second round.

PepMS commented:
**RESULTS:** (images attached at the end)
humanoid_taichy.py
```
******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.5, running with linear solver ma27.

Number of nonzeros in equality constraint Jacobian...:  1683856
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:   709246

Total number of variables............................:    13036
                     variables with only lower bounds:        0
                variables with lower and upper bounds:     3840
                     variables with only upper bounds:        0
Total number of equality constraints.................:     9196
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  1.7899356e+02 9.09e-14 1.73e+00   0.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  1.3016989e+02 1.61e-01 1.06e+01   0.6 4.08e+00    -  9.96e-01 2.50e-01f  3
   2  1.1781888e+02 1.58e-01 9.33e+01  -0.3 1.22e+01    -  9.97e-01 1.25e-01f  4
   3  9.8967162e+01 1.74e-01 8.16e+01  -0.8 4.80e+01    -  9.98e-01 1.25e-01f  4
   4  8.3449708e+01 3.41e-01 7.14e+01  -1.2 1.08e+02    -  9.94e-01 1.25e-01f  4
   5  7.9508035e+01 8.20e-01 1.06e+02  -1.4 5.60e+01    -  1.00e+00 1.25e-01f  4
   6  7.6969386e+01 8.91e-01 9.23e+01  -1.7 5.43e+01    -  1.00e+00 1.25e-01f  4
   7  6.5738503e+01 1.67e+00 6.88e+01  -1.8 6.14e+01    -  1.00e+00 2.50e-01f  3
   8  4.6502041e+01 5.08e+00 5.12e+01  -1.9 4.15e+01    -  1.00e+00 2.50e-01f  3
   9  2.8779443e+01 4.91e+00 5.36e+01  -2.3 3.33e+01    -  1.00e+00 5.00e-01f  2
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  10  1.7176791e+01 5.69e+00 8.91e+01  -2.0 3.30e+01    -  9.96e-01 1.00e+00f  1
  11  8.0712215e+00 2.90e+00 6.63e+00  -2.2 3.88e+01    -  1.00e+00 1.00e+00h  1
  12  7.2921188e+00 3.27e+00 3.56e+00  -2.5 1.84e+01    -  9.99e-01 1.00e+00h  1
  13  1.0035186e+01 6.89e-01 4.97e+00  -3.0 1.14e+01    -  9.99e-01 1.00e+00h  1
  14  7.0917853e+00 4.84e-01 1.18e+01  -3.7 5.19e+00    -  1.00e+00 1.00e+00h  1
  15  7.0302604e+00 2.80e+00 6.35e+00  -4.3 7.84e+00    -  1.00e+00 5.00e-01h  2
  16  7.5078644e+00 1.99e+00 7.74e+00  -4.0 5.06e+00    -  1.00e+00 5.00e-01h  2
  17  6.8563730e+00 1.54e+00 8.07e-01  -4.1 4.24e+00    -  1.00e+00 1.00e+00h  1
  18  6.8512730e+00 9.35e-01 9.00e-01  -4.7 3.03e+00    -  1.00e+00 1.00e+00h  1
  19  6.8470712e+00 9.78e-01 1.22e+00  -5.4 2.86e+00    -  1.00e+00 2.50e-01h  3
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  20  6.8720733e+00 1.02e+00 1.80e+00  -6.2 2.99e+00    -  1.00e+00 5.00e-01h  2
  21  6.8601804e+00 3.23e-01 1.53e+00  -7.1 2.59e+00    -  1.00e+00 1.00e+00h  1
  22  6.8292449e+00 1.13e+00 6.21e-01  -6.6 2.09e+00    -  1.00e+00 1.00e+00h  1
  23  6.8347083e+00 7.25e-01 1.15e+00  -6.8 1.78e+00    -  1.00e+00 5.00e-01h  2
  24  6.8206472e+00 3.80e-01 3.14e-01  -7.6 1.79e+00    -  1.00e+00 1.00e+00h  1
  25  6.8232665e+00 2.06e-02 5.75e-01  -8.6 1.43e+00    -  1.00e+00 1.00e+00H  1
  26  6.8191299e+00 7.44e-02 1.83e-01  -8.9 1.17e+00    -  1.00e+00 1.00e+00h  1
  27  6.8184533e+00 7.56e-03 1.15e-02 -10.0 1.00e+00    -  1.00e+00 1.00e+00h  1
  28  6.8184554e+00 3.93e-04 2.06e-02 -11.0 8.01e-01    -  1.00e+00 1.00e+00H  1
  29  6.8184394e+00 3.92e-04 4.76e-03 -11.0 6.42e-01    -  1.00e+00 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  30  6.8184341e+00 2.04e-04 3.48e-03 -11.0 4.44e-01    -  1.00e+00 1.00e+00h  1
  31  6.8184314e+00 1.04e-04 1.78e-03 -11.0 2.92e-01    -  1.00e+00 1.00e+00h  1
  32  6.8184266e+00 1.85e-04 4.23e-04 -11.0 1.78e-01    -  1.00e+00 1.00e+00h  1
  33  6.8184263e+00 1.64e-06 3.55e-04 -11.0 7.87e-03    -  1.00e+00 1.00e+00h  1
  34  6.8184262e+00 1.00e-06 2.26e-04 -11.0 4.14e-03    -  1.00e+00 1.00e+00h  1
  35  6.8184262e+00 4.50e-07 1.90e-04 -11.0 3.45e-03    -  1.00e+00 1.00e+00h  1
  36  6.8184261e+00 2.75e-07 1.20e-04 -11.0 2.29e-03    -  1.00e+00 1.00e+00h  1
  37  6.8184261e+00 1.22e-07 1.04e-04 -11.0 1.92e-03    -  1.00e+00 1.00e+00h  1
  38  6.8184261e+00 7.69e-08 6.34e-05 -11.0 1.28e-03    -  1.00e+00 1.00e+00h  1
  39  6.8184261e+00 3.59e-08 5.70e-05 -11.0 1.09e-03    -  1.00e+00 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  40  6.8184261e+00 2.26e-08 3.34e-05 -11.0 7.24e-04    -  1.00e+00 1.00e+00h  1
  41  6.8184261e+00 1.10e-08 3.14e-05 -11.0 6.19e-04    -  1.00e+00 1.00e+00h  1
  42  6.8184261e+00 6.97e-09 1.76e-05 -11.0 4.11e-04    -  1.00e+00 1.00e+00h  1
  43  6.8184261e+00 3.44e-09 1.74e-05 -11.0 3.55e-04    -  1.00e+00 1.00e+00h  1
  44  6.8184261e+00 2.19e-09 9.23e-06 -11.0 2.35e-04    -  1.00e+00 1.00e+00h  1
  45  6.8184261e+00 1.10e-09 9.77e-06 -11.0 2.06e-04    -  1.00e+00 1.00e+00h  1
  46  6.8184261e+00 7.04e-10 4.84e-06 -11.0 1.37e-04    -  1.00e+00 1.00e+00h  1
  47  6.8184261e+00 3.59e-10 5.54e-06 -11.0 1.21e-04    -  1.00e+00 1.00e+00h  1
  48  6.8184261e+00 2.33e-10 2.60e-06 -11.0 8.04e-05    -  1.00e+00 1.00e+00h  1
  49  6.8184261e+00 1.21e-10 3.18e-06 -11.0 7.21e-05    -  1.00e+00 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  50  6.8184261e+00 7.92e-11 1.46e-06 -11.0 4.80e-05    -  1.00e+00 1.00e+00h  1
  51  6.8184261e+00 4.16e-11 1.85e-06 -11.0 4.37e-05    -  1.00e+00 1.00e+00h  1
  52  6.8184261e+00 2.77e-11 8.29e-07 -11.0 2.91e-05    -  1.00e+00 1.00e+00h  1
  53  6.8184261e+00 1.48e-11 1.09e-06 -11.0 2.68e-05    -  1.00e+00 1.00e+00h  1
  54  6.8184261e+00 9.96e-12 4.77e-07 -11.0 1.79e-05    -  1.00e+00 1.00e+00h  1
  55  6.8184261e+00 5.41e-12 6.55e-07 -11.0 1.67e-05    -  1.00e+00 1.00e+00h  1
  56  6.8184261e+00 3.75e-12 2.79e-07 -11.0 1.11e-05    -  1.00e+00 1.00e+00h  1
  57  6.8184261e+00 2.09e-12 3.99e-07 -11.0 1.05e-05    -  1.00e+00 1.00e+00h  1
  58  6.8184261e+00 1.40e-12 1.66e-07 -11.0 7.02e-06    -  1.00e+00 1.00e+00h  1
  59  6.8184261e+00 9.00e-13 2.46e-07 -11.0 6.72e-06    -  1.00e+00 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  60  6.8184261e+00 5.89e-13 1.04e-07 -11.0 4.48e-06    -  1.00e+00 1.00e+00h  1
  61  6.8184261e+00 3.54e-13 1.54e-07 -11.0 4.33e-06    -  1.00e+00 1.00e+00h  1
  62  6.8184261e+00 2.86e-13 6.58e-08 -11.0 3.01e-06    -  1.00e+00 1.00e+00h  1

Number of Iterations....: 62

                                   (scaled)                 (unscaled)
Objective...............:   6.8184261112987965e+00    6.8184261112987965e+00
Dual infeasibility......:   6.5840106255876663e-08    6.5840106255876663e-08
Constraint violation....:   2.8643754035329039e-13    2.8643754035329039e-13
Variable bound violation:   2.2161226809203072e-07    2.2161226809203072e-07
Complementarity.........:   1.0000020834585570e-11    1.0000020834585570e-11
Overall NLP error.......:   6.5840106255876663e-08    6.5840106255876663e-08


Number of objective function evaluations             = 99
Number of objective gradient evaluations             = 63
Number of equality constraint evaluations            = 108
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 63
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 62
Total seconds in IPOPT                               = 35.346

EXIT: Optimal Solution Found.
Finally reached = (0.400, 0.000, 1.200)
Distance between hand and target = 4.677E-05
Distance to default state = 1.605E+00
XY distance to CoM reference = 9.818E-02
```
quadrotor_ubound.py
```

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.5, running with linear solver ma27.

Number of nonzeros in equality constraint Jacobian...:    11232
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:     4566

Total number of variables............................:      540
                     variables with only lower bounds:        0
                variables with lower and upper bounds:      132
                     variables with only upper bounds:        0
Total number of equality constraints.................:      408
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  3.0099000e+00 2.85e-01 9.71e-02   0.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  2.2932406e+01 1.22e-02 3.96e+00   0.2 5.46e+00    -  9.33e-01 1.00e+00f  1
   2  1.2893510e+01 2.69e-01 1.21e+00  -0.5 4.01e+00    -  9.24e-01 1.00e+00f  1
   3  4.1037796e+00 6.96e-01 1.70e+00  -6.3 5.52e+00    -  4.73e-01 1.00e+00f  1
   4  1.1880429e+00 8.88e-01 9.00e-01  -0.7 7.01e+00    -  1.00e+00 1.00e+00f  1
   5  3.8821112e-01 3.95e-01 7.87e-01  -6.7 5.07e+00    -  5.84e-01 1.00e+00h  1
   6  2.6122153e-01 2.20e-01 6.28e-02  -2.4 4.27e+00    -  5.97e-01 6.55e-01h  1
   7  2.0304810e-01 1.18e-01 6.82e-02  -2.5 2.83e+00    -  7.76e-01 8.62e-01h  1
   8  2.7537296e-02 3.23e-02 1.62e-02  -8.0 1.58e+00    -  3.48e-01 1.00e+00h  1
   9  1.2045705e-02 1.27e-03 1.40e-03  -3.5 2.85e-01    -  9.98e-01 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  10  7.3235001e-03 7.70e-04 2.10e-04  -3.9 3.01e-01    -  9.92e-01 1.00e+00h  1
  11  7.0189944e-03 1.90e-03 7.65e-05  -5.3 7.51e-01    -  9.27e-01 1.00e+00h  1
  12  6.8255427e-03 1.07e-03 9.55e-05  -5.4 1.02e+00    -  9.91e-01 1.00e+00h  1
  13  6.6080303e-03 6.57e-04 3.75e-05  -5.8 1.23e+00    -  9.90e-01 1.00e+00h  1
  14  6.5249614e-03 2.37e-04 6.82e-06  -6.3 9.52e-01    -  9.83e-01 1.00e+00h  1
  15  6.4939680e-03 7.15e-05 7.38e-06  -6.8 9.10e-01    -  9.94e-01 1.00e+00h  1
  16  6.4837533e-03 5.15e-05 5.79e-06  -7.5 5.89e-01    -  9.94e-01 1.00e+00h  1
  17  6.4819157e-03 1.24e-05 2.47e-06  -8.0 3.30e-01    -  1.00e+00 1.00e+00h  1
  18  6.4815478e-03 2.15e-06 8.13e-07  -8.2 1.88e-01    -  9.97e-01 1.00e+00h  1
  19  6.4812087e-03 7.78e-07 9.93e-07  -9.1 1.37e-01    -  1.00e+00 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  20  6.4811825e-03 1.14e-07 3.66e-07  -9.3 5.64e-02    -  1.00e+00 1.00e+00h  1
  21  6.4811500e-03 4.39e-08 2.76e-07 -10.1 2.38e-02    -  1.00e+00 1.00e+00h  1
  22  6.4811441e-03 1.28e-08 1.48e-07 -11.0 6.97e-03    -  1.00e+00 1.00e+00h  1
  23  6.4811438e-03 8.62e-10 5.35e-08 -11.0 2.39e-03    -  1.00e+00 1.00e+00h  1
  24  6.4811437e-03 3.99e-11 1.17e-08 -11.0 6.03e-04    -  1.00e+00 1.00e+00h  1
  25  6.4811437e-03 4.66e-13 7.99e-10 -11.0 6.51e-05    -  1.00e+00 1.00e+00h  1

Number of Iterations....: 25

                                   (scaled)                 (unscaled)
Objective...............:   6.4811437277397591e-03    6.4811437277397591e-03
Dual infeasibility......:   7.9929055521466426e-10    7.9929055521466426e-10
Constraint violation....:   4.6613927995320381e-13    4.6613927995320381e-13
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   1.0010916266226702e-11    1.0010916266226702e-11
Overall NLP error.......:   7.9929055521466426e-10    7.9929055521466426e-10


Number of objective function evaluations             = 26
Number of objective gradient evaluations             = 26
Number of equality constraint evaluations            = 26
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 26
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 25
Total seconds in IPOPT                               = 0.116

EXIT: Optimal Solution Found.
```
arm_manipulator.py
```

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.5, running with linear solver ma27.

Number of nonzeros in equality constraint Jacobian...:   122696
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:    57855

Total number of variables............................:     5264
                     variables with only lower bounds:        0
                variables with lower and upper bounds:     1750
                     variables with only upper bounds:        0
Total number of equality constraints.................:     3514
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  6.5678440e-01 1.00e+00 1.36e-05   0.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  1.3266754e+00 6.54e-03 2.11e-01  -1.1 2.56e+00    -  9.94e-01 1.00e+00h  1
   2  1.1094118e+00 6.45e-04 3.01e-02  -6.8 1.39e+00    -  9.28e-01 1.00e+00h  1
   3  7.6713684e-01 5.14e-03 4.97e-02  -3.1 4.01e+00    -  1.00e+00 1.00e+00f  1
   4  7.3549218e-01 1.19e-03 1.41e-02  -4.1 2.61e+00    -  9.69e-01 1.00e+00h  1
   5  6.2755204e-01 8.50e-03 1.52e-01  -4.4 1.31e+01    -  1.00e+00 6.60e-01h  1
   6  6.5493204e-01 2.40e-03 6.87e-02  -3.5 1.61e+01    -  1.00e+00 9.33e-01H  1
   7  6.1728209e-01 1.95e-02 8.93e-02  -3.8 7.03e+00    -  9.95e-01 1.00e+00h  1
   8  5.8405495e-01 1.05e-02 4.75e-02  -4.4 4.45e+00    -  9.96e-01 1.00e+00h  1
   9  5.4881583e-01 6.15e-03 9.32e-02  -4.5 6.01e+00    -  1.00e+00 9.99e-01h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  10  5.3830202e-01 4.52e-03 2.65e-02  -4.6 7.63e+00    -  1.00e+00 1.00e+00h  1
  11  5.2060600e-01 2.42e-02 6.34e-02  -5.2 6.83e+00    -  9.99e-01 1.00e+00h  1
  12  5.2187001e-01 2.90e-02 7.81e-02  -4.3 1.81e+01    -  1.00e+00 1.97e-01h  1
  13  5.0903387e-01 1.39e-01 2.23e-01  -4.5 1.37e+01    -  7.85e-01 5.11e-01h  1
  14  4.5924108e-01 2.22e-01 3.14e-01  -4.1 2.54e+01    -  1.00e+00 6.51e-01h  1
  15  3.8921836e-01 1.61e-01 1.09e-01  -4.3 2.45e+01    -  4.18e-01 8.44e-01h  1
  16  3.6568354e-01 8.20e-02 6.32e-02  -4.4 3.15e+01    -  6.38e-01 6.02e-01h  1
  17  3.4163412e-01 5.46e-02 5.37e-02  -4.7 3.35e+01    -  6.18e-01 5.22e-01h  1
  18  3.2996835e-01 3.41e-02 4.01e-02  -4.6 3.58e+01    -  9.14e-01 3.93e-01h  1
  19  3.3049634e-01 1.32e-02 1.97e-02  -4.4 2.66e+01    -  1.00e+00 6.22e-01h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  20  3.1585887e-01 7.81e-03 8.19e-03  -4.8 4.38e+00    -  1.00e+00 1.00e+00h  1
  21  3.0442952e-01 4.29e-03 6.18e-03  -5.3 7.35e+00    -  1.00e+00 1.00e+00h  1
  22  2.9954527e-01 1.82e-03 4.92e-03  -5.6 6.57e+00    -  1.00e+00 1.00e+00h  1
  23  2.9855921e-01 5.59e-04 2.98e-03  -5.8 3.57e+00    -  9.40e-01 8.59e-01h  1
  24  2.9764920e-01 3.14e-04 2.17e-03  -6.0 3.02e+00    -  9.87e-01 8.12e-01h  1
  25  2.9709906e-01 1.79e-04 1.67e-03  -6.2 3.58e+00    -  9.98e-01 7.98e-01h  1
  26  2.9698531e-01 7.75e-05 1.34e-03  -6.2 3.65e+00    -  1.00e+00 9.20e-01h  1
  27  2.9646325e-01 7.95e-05 8.48e-04  -6.6 1.85e+00    -  1.00e+00 1.00e+00h  1
  28  2.9631839e-01 4.91e-05 7.53e-04  -6.8 2.16e+00    -  1.00e+00 1.00e+00h  1
  29  2.9621446e-01 2.49e-05 4.91e-04  -7.1 1.65e+00    -  1.00e+00 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  30  2.9617255e-01 1.12e-05 3.65e-04  -7.3 1.12e+00    -  1.00e+00 9.95e-01h  1
  31  2.9612491e-01 5.90e-06 2.36e-04  -7.8 7.12e-01    -  1.00e+00 1.00e+00h  1
  32  2.9612106e-01 1.88e-06 1.78e-04  -7.9 5.19e-01    -  1.00e+00 1.00e+00h  1
  33  2.9610728e-01 1.08e-06 1.07e-04  -8.9 3.13e-01    -  1.00e+00 1.00e+00h  1
  34  2.9610625e-01 3.40e-07 7.46e-05  -9.4 2.30e-01    -  1.00e+00 1.00e+00h  1
  35  2.9610575e-01 1.24e-07 4.47e-05 -10.1 1.35e-01    -  1.00e+00 1.00e+00h  1
  36  2.9610566e-01 4.84e-08 2.78e-05 -11.0 8.47e-02    -  1.00e+00 1.00e+00h  1
  37  2.9610566e-01 1.62e-08 1.68e-05 -11.0 5.05e-02    -  1.00e+00 1.00e+00h  1
  38  2.9610566e-01 6.43e-09 1.02e-05 -11.0 3.09e-02    -  1.00e+00 1.00e+00h  1
  39  2.9610566e-01 2.19e-09 6.15e-06 -11.0 1.85e-02    -  1.00e+00 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  40  2.9610566e-01 8.55e-10 3.74e-06 -11.0 1.13e-02    -  1.00e+00 1.00e+00h  1
  41  2.9610566e-01 2.95e-10 2.25e-06 -11.0 6.78e-03    -  1.00e+00 1.00e+00h  1
  42  2.9610566e-01 1.14e-10 1.37e-06 -11.0 4.14e-03    -  1.00e+00 1.00e+00h  1
  43  2.9610566e-01 3.98e-11 8.25e-07 -11.0 2.48e-03    -  1.00e+00 1.00e+00h  1
  44  2.9610566e-01 1.52e-11 5.01e-07 -11.0 1.51e-03    -  1.00e+00 1.00e+00h  1
  45  2.9610566e-01 5.35e-12 3.02e-07 -11.0 9.10e-04    -  1.00e+00 1.00e+00h  1
  46  2.9610566e-01 2.04e-12 1.83e-07 -11.0 5.54e-04    -  1.00e+00 1.00e+00h  1
  47  2.9610566e-01 7.14e-13 1.11e-07 -11.0 3.33e-04    -  1.00e+00 1.00e+00h  1
  48  2.9610566e-01 2.74e-13 6.71e-08 -11.0 2.03e-04    -  1.00e+00 1.00e+00h  1
  49  2.9610566e-01 9.24e-14 4.05e-08 -11.0 1.22e-04    -  1.00e+00 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  50  2.9610566e-01 3.91e-14 2.46e-08 -11.0 7.42e-05    -  1.00e+00 1.00e+00h  1
  51  2.9610566e-01 1.42e-14 1.48e-08 -11.0 4.47e-05    -  1.00e+00 1.00e+00h  1
  52  2.9610566e-01 7.11e-15 8.99e-09 -11.0 2.71e-05    -  1.00e+00 1.00e+00h  1
  53  2.9610566e-01 3.55e-15 5.43e-09 -11.0 1.64e-05    -  1.00e+00 1.00e+00h  1
  54  2.9610566e-01 3.55e-15 3.29e-09 -11.0 9.93e-06    -  1.00e+00 1.00e+00h  1
  55  2.9610566e-01 3.55e-15 1.99e-09 -11.0 6.00e-06    -  1.00e+00 1.00e+00h  1
  56  2.9610566e-01 3.55e-15 1.21e-09 -11.0 3.64e-06    -  1.00e+00 1.00e+00h  1
  57  2.9610566e-01 3.55e-15 7.28e-10 -11.0 2.20e-06    -  1.00e+00 1.00e+00h  1

Number of Iterations....: 57

                                   (scaled)                 (unscaled)
Objective...............:   2.9610565990200172e-01    2.9610565990200172e-01
Dual infeasibility......:   7.2830866337803002e-10    7.2830866337803002e-10
Constraint violation....:   3.5527136788005009e-15    3.5527136788005009e-15
Variable bound violation:   3.9502271675928569e-08    3.9502271675928569e-08
Complementarity.........:   1.0000000114926333e-11    1.0000000114926333e-11
Overall NLP error.......:   7.2830866337803002e-10    7.2830866337803002e-10


Number of objective function evaluations             = 59
Number of objective gradient evaluations             = 58
Number of equality constraint evaluations            = 59
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 58
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 57
Total seconds in IPOPT                               = 2.367

EXIT: Optimal Solution Found.
```
![ipopt_taichy](https://user-images.githubusercontent.com/38432035/171565999-500d460b-2a51-4b8b-ac63-370d750e2088.png)
![ipopt_quadrotor_ubound](https://user-images.githubusercontent.com/38432035/171565997-bd8527a9-793e-4a37-9a48-99393940ab87.png)
![ipopt_arm_manipulator](https://user-images.githubusercontent.com/38432035/171565995-32c0412f-94b2-4682-bc15-d578bbba4b13.png)

cmastalli commented:
> I'm not comfortable doing otherwise. Unfortunately, right now I don't have the time to do it.

I understand. Let's do this optimisation later.

cmastalli reviewed (approved):
@PepMS 

Thank you so much for this important effort.
On top of your PR, I have made substantial changes:
 1. Enable multithreading in Ipopt routines
 2. Removed warming during compilation
 3. Enable resizeData used in MPC (I haven't tested, but let's keep for later)

It is missing to work on the optimisation discussed and to integrate some examples.
For now, I think we could go without those points.

cmastalli commented:
@PepMS  -- last bit. Could you test the solver with multithreading support? My tests shows a correct behaviour, but I would like to double check that this is happening from your side as well.

wxmerkt commented:
@cmastalli I've added a commit which should enable IPOPT for CI

cmastalli commented:
> @cmastalli I've added a commit which should enable IPOPT for CI

Thank you so much ;)

cmastalli commented:
@wxmerkt do you have idea why the ROS jobs are not passing?

PepMS commented:
> @wxmerkt do you have idea why the ROS jobs are not passing?

I think it has something to do with this: https://github.com/casadi/casadi/issues/1010

PepMS commented:
I've done what the link said and it works. However, this is a workaround. I've tested this with the version from apt, which is the 3.11. With the one compiled from source, which is the 3.14 this is not needed.

In any case the tests pass now.
